app/controllers/lti_nrps_controller.rb

```diff
@@ -0,0 +1,132 @@
+  skip_before_action :update_persistent_announcements
+
+  rescue_from LtiLaunchController::LtiError, with: :respond_with_lti_error
+  def respond_with_lti_error(error)
+    Rails.logger.send(:warn) { "Lti NRPS Error: #{error.message}" }
+    render json: { error: error.message }.to_json, status: :bad_request
+  end
+  action_auth_level :request_access_token, :instructor
+  def request_access_token
+    # get private key from config to sign Autolab's client assertion as a JWK
+    private_key = Rails.configuration.lti_settings["tool_private_key"].to_s.gsub(/\\n/, "\n")
+    tool_rsa_private = OpenSSL::PKey::RSA.new(private_key)
+    # build JWK format for RSA
+    optional_parameters = {
+      kid: "dGkeZwrt+H7GnGZ2t2LUfhYf+/o=",
+      use: 'sig',
+      alg: 'RS256',
+      e: "AQAB",
+      kty: "RSA",
+    }
+    tool_rsa_private_JWK = JWT::JWK.new(tool_rsa_private, optional_parameters);
+    # build client assertion based on lti 1.3 spec
+    # https://www.imsglobal.org/spec/security/v1p0/#using-json-web-tokens-with-oauth-2-0-client-credentials-grant
+    # https://www.imsglobal.org/spec/lti/v1p3#token-endpoint-claim-and-services
+    client_assertion = {
+      "iss": Rails.configuration.lti_settings["developer_key"],
+      "sub": Rails.configuration.lti_settings["developer_key"],
+      "aud": Rails.configuration.lti_settings["platform_oauth2_access_token_url"],
+      "iat": Time.now.to_i,
+      "exp": Time.now.to_i + 600,
+      "jti": "lti-refresh-token-#{SecureRandom.uuid}"
+    }
+    # sign client_assertion using private key
+    token = JWT.encode(client_assertion, tool_rsa_private_JWK.keypair, optional_parameters[:alg],
+                       kid: optional_parameters[:kid])
+    # build Client-Credentials Grant
+    # https://www.imsglobal.org/spec/security/v1p0/#using-oauth-2-0-client-credentials-grant
+    payload = {
+      "grant_type": "client_credentials",
+      "client_assertion_type": "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
+      "client_assertion": token,
+      "scope": "https://purl.imsglobal.org/spec/lti-nrps/scope/contextmembership.readonly"
+    }
+    # send Client-Credentials Grant to LTI Oauth2 access token endpoint
+    conn = Faraday.new(
+      url: Rails.configuration.lti_settings["platform_oauth2_access_token_url"],
+      headers: { 'Content-Type' => 'application/json' }
+    )
+    response = conn.post('') do |req|
+      req.body = payload.to_json
+    end
+    response_body = JSON.parse(response.body)
+    if response_body["access_token"].nil?
+      raise LtiLaunchController::LtiError.new("Client-Credentials Grant Failed: #{response.body}",
+                                              :bad_request)
+    end
+
+    response_body["access_token"]
+  end
+
+  # NRPS endpoint for Autolab to send an NRPS request to LTI Advantage Platform
+  action_auth_level :send_nrps_request, :instructor
+  def send_nrps_request
+    params.require(:lcd_id)
```

Comment:
Nit: is this check required since you validate `lcd.nil?` later?

Regardless, it's better to check for presence than to use `require`, see: https://stackoverflow.com/questions/41869614/how-to-check-for-parameter-presence-params-requiredata-or-paramsdata-pres

Same comment for users_controllers `lti_launch_initialize` and `lti_launch_link_course `

Reply:
Check isn't required, will remove it.